Math/NumberTheory/Moduli/CubicSymbol.hs

```diff
@@ -0,0 +1,125 @@
+{-# LANGUAGE LambdaCase #-}
+{-# LANGUAGE DataKinds #-}
+
+module Math.NumberTheory.Moduli.CubicSymbol
+  ( CubicSymbol(..)
+  , cubicSymbol
+  ) where
+
+import Math.NumberTheory.Quadratic.EisensteinIntegers
+import Math.NumberTheory.Utils.FromIntegral
+import qualified Data.Euclidean as A
+import Math.NumberTheory.Utils
+import Data.Semigroup
+import Data.Mod.Word
+import Data.Maybe
+import Data.List
+
+data CubicSymbol = Zero | Omega | OmegaSquare | One deriving (Eq)
+
+instance Semigroup CubicSymbol where
+  Zero <> _                    = Zero
+  _ <> Zero                    = Zero
+  One <> y                     = y
+  x <> One                     = x
+  Omega <> Omega               = OmegaSquare
+  Omega <> OmegaSquare         = One
+  OmegaSquare <> Omega         = One
+  OmegaSquare <> OmegaSquare   = Omega
+
+instance Show CubicSymbol where
+  show = \case
+    Zero         -> "0"
+    Omega        -> "ω"
+    OmegaSquare  -> "ω²"
+    One          -> "1"
+
+-- The algorithm cubicSymbol takes two Eisentein numbers @alpha@ and @beta@ and returns
+-- their cubic residue. It is divided in the following steps.
+
+-- 1) Check whether @beta@ is coprime to 3.
+-- 2) Replace @alpha@ by the remainder of @alpha@ mod @beta@
+--    This does not affect the cubic symbol.
+-- 3) Replace @alpha@ and @beta@ by their associated primary
+--    divisors and keep track of how their cubic residue changes.
+-- 4) Check if any of the two numbers is a zero or a unit. If it
+--    is, return their cubic residue.
+-- 5) If not, invoke cubic reciprocity by swapping @alpha@ and
+--    @beta@. Note both numbers have to be primary.
+--    Return to Step 2.
+
+-- This function takes two Eisenstein integers and returns their cubic residue character.
+-- Note that the second argument must be coprime to 3 else the algorithm returns an error.
+cubicSymbol :: EisensteinInteger -> EisensteinInteger -> CubicSymbol
+cubicSymbol alpha beta = case betaNorm `mod` 3 of
+  -- This checks whether beta is coprime to 3, i.e. divisible by @1 - ω@
+  -- In particular, it returns an error if @beta == 0@
+  0 -> error "Math.NumberTheory.Moduli.CubicSymbol: denominator is not coprime to 3."
+  _ -> cubicSymbolHelper alpha beta
+  where
+    betaNorm = norm beta
+
+cubicSymbolHelper :: EisensteinInteger -> EisensteinInteger -> CubicSymbol
+cubicSymbolHelper alpha beta = cubicReciprocity primaryRemainder primaryBeta <> newSymbol
```

Comment:
I suggest to inline this definition into `cubicSymbol`. These function are not large, not much point to introduce a new name and an additional indirection.

Reply:
Sure. The only reason why I didn't do it this way was to avoid checking multiple times if 3 divides the norm.

Reply:
Ah, that's a good point. Then leave it as is.